Docker/Image.php

```diff
@@ -151,25 +156,43 @@ public function getSourceComponent()
     }
 
     /**
-     * Get and log hash of the image.
-     * @param string $name Image name including tag
+     * Log repository hash (digest) of the image.
+    */
+    public function logImageHash()
+    {
+        $this->logger->notice(
+            "Using image " . $this->getFullImageId() . " with repo-digest " .
+            implode(', ', $this->getImageDigests())
+        );
+    }
+
+    /**
+     * Get repository hash (digest) of the image.
+     * @return string[]
      */
-    public function logImageHash($name)
+    public function getImageDigests()
     {
-        $retryPolicy = new SimpleRetryPolicy(3);
-        $backOffPolicy = new ExponentialBackOffPolicy(10000);
-        $proxy = new RetryProxy($retryPolicy, $backOffPolicy);
-        $command = "sudo docker images --format \"{{.ID}}\" --no-trunc " . escapeshellarg($name);
-
-        $process = new Process($command);
-        $process->setTimeout(3600);
-        try {
-            $proxy->call(function () use ($process) {
-                $process->mustRun();
-            });
-            $this->logger->notice("Using image $name with hash " . trim($process->getOutput()));
-        } catch (\Exception $e) {
-            $this->logger->error("Failed to get hash for image " . $name);
+        if (empty($this->imageDigests)) {
+            $command = "sudo docker inspect " . escapeshellarg($this->getFullImageId());
+            $process = new Process($command);
+            $process->setTimeout(3600);
+            try {
+                $retryPolicy = new SimpleRetryPolicy(3);
+                $backOffPolicy = new ExponentialBackOffPolicy(10000);
+                $proxy = new RetryProxy($retryPolicy, $backOffPolicy);
+                $proxy->call(function () use ($process) {
+                    $process->mustRun();
+                });
+                $inspect = json_decode($process->getOutput(), true);
+                if ((json_last_error() != JSON_ERROR_NONE) && !empty($inspect[0]['RepoDigests'])) {
+                    throw new \InvalidArgumentException("Inspect error " . json_last_error_msg());
+                }
+                $this->imageDigests = $inspect[0]['RepoDigests'];
```

Comment:
`RepoDigests` funguje na ECR, ale pro quayio je to prázdný (DockerHub jsem nezkoušel). Mám to něčím nahradit, třeba `Id`?

![image](https://user-images.githubusercontent.com/497675/29541458-03341b00-8689-11e7-93cc-4156c0fd3a8a.png)

Reply:
to je divny, normalne to funguje i pro quay https://github.com/keboola/docker-bundle/pull/187/files#diff-01b7efa37de70021fb2429bb1d8a4640R304 a zkousel jsem i quay.io/keboola/docker-demo-app:1.1.6

Docker version 17.06.0-ce, build 02c1d87

Reply:
```
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   b9f10c9/1.11.2
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   b9f10c9/1.11.2
 Built:        
 OS/Arch:      linux/amd64
```

Reply:
Dělám to na devel-03. jak to testuješ/vyvíjíš? Lokálně? Chtěl jsem udělat end-to-end test, založil jsem si job a zkoumám, co je v resultu.

```
{
  "id": 315053412,
  "runId": "315053413",
  "lockName": "docker-395-docker-demo-quayio-59adcd5aa77c9",
  "project": {
    "id": 395,
    "name": "[Test] Syrup DEVEL"
  },
  "token": {
    "id": "11024",
    "description": "ondrej.hlavacek@keboola.com"
  },
  "component": "docker",
  "command": "run",
  "params": {
    "configData": {
      "parameters": {
        "data_column": "text",
        "primary_key_column": "id",
        "string_length": "255"
      },
      "storage": {
        "input": {
          "tables": [
            {
              "source": "in.c-docker-demo.data"
            }
          ]
        },
        "output": {
          "tables": [
            {
              "source": "sliced.csv",
              "destination": "out.c-docker-demo.data"
            }
          ]
        }
      }
    },
    "component": "docker-demo-quayio",
    "mode": "run"
  },
  "result": {
    "message": "Component processing finished.",
    "images": [
      {
        "id": "quay.io\/keboola\/docker-demo-app:1.1.6",
        "digests": []
      }
    ]
  },
  "status": "success",
  "process": {
    "host": "kbc-devel-03",
    "pid": 10595
  },
  "createdTime": "2017-09-05T00:02:02+02:00",
  "startTime": "2017-09-05T00:02:14+02:00",
  "endTime": "2017-09-05T00:02:30+02:00",
  "durationSeconds": 16,
  "waitSeconds": 12,
  "nestingLevel": 0,
  "error": null,
  "errorNote": null,
  "terminatedBy": {
    "id": null,
    "description": null
  },
  "encrypted": null,
  "usage": [],
  "_index": "devel_syrup_docker_2017_3",
  "_type": "jobs",
  "isFinished": true
}
```

Reply:
no tak to mas asi starej docker, delam to lokalne a na jenkinsovi je test, normalne prochazi https://jenkins.keboola.com/view/Docker%20Runner/job/PHPUnit%20-%20docker-bundle%20(branch)/ - na devel2 je 17 a na produkci je tak 17

Reply:
oukej, ještě to zkusím na devel2!

Reply:
potvrzuju, na devel-02 to šlape!

Reply:
tak asi merge! :-)

Reply:
ouk, po obede na to vlitnu, Ď